byron-proxy/src/lib/Ouroboros/Byron/Proxy/Server/Socket.hs

```diff
@@ -0,0 +1,248 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE RecursiveDo #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE BangPatterns #-}
+
+module Ouroboros.Byron.Proxy.Server.Socket
```

Comment:
It let's the user decide whether to kill all running threads or to wait for them to stop on their own. In our production application, we'll want to kill them, because they are expected to go on forever. But if you were using this server to support, say, something like an HTTP web server, you'd probably prefer to wait for them all to stop, since they're expected to be fairly quick and short-lived.